File: problems/1-easy/110-Balanced-Binary-Tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.isbalanced = True
        self.dfs(root)
        return self.isbalanced

    def dfs(self, root):
        if not root:
            return 0

        left_height = self.dfs(root.left)

        right_height = self.dfs(root.right)

        if abs(left_height - right_height) > 1:
            self.isbalanced = False

        return max(left_height, right_height) + 1

File: problems/1-easy/110-Balanced-Binary-Tree/solution.cpp
#include <algorithm>
#include <cmath>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int balanced = true;
    bool isBalanced(TreeNode* root) {
        balanced = true;
        dfs(root);
        return balanced;
    }

    int dfs(TreeNode* root) {
        if (!balanced || root == nullptr) {
            return 0;
        }

        int left = dfs(root->left);
        int right = dfs(root->right);

        if (abs(left-right) > 1) {
            balanced = false;
        }
        return max(left, right) + 1;
    }
};